tempesta_fw/msg.c

```diff
@@ -34,20 +34,22 @@
 int
 tfw_msg_write(TfwMsgIter *it, const TfwStr *data)
 {
-	char *p;
 	const TfwStr *c, *end;
-	skb_frag_t *frag;
-	unsigned int c_off = 0, f_size, c_size, f_room, n_copy;
 
 	BUG_ON(TFW_STR_DUP(data));
+	if (WARN_ON_ONCE(it->frag >= skb_shinfo(it->skb)->nr_frags))
+		return -E2BIG;
```

Comment:
In case of `*skb == skb_head` in `ss_skb_frag_next()` we get here `skb_head == it->skb` and `0 == it->frag`; so, condition `it->frag >= skb_shinfo(it->skb)->nr_frags)` probably will be false here and we continue execute all the following `tfw_msg_write()` calls without warnings.

Reply:
The scenario you've described  may happen on subsequent `tfw_msg_write()` calls, which is very common case. So yes, the condition will false and we will continue to execute the function. But `*skb == skb_head` really means that we run out of the spare place in the skb fragments, and no more data can be pushed there. So if we continue to execute the function, we will walk through all the fragments once again and return `-E2BIG` error.

So there is no errors and the function can be called multiple times without any worries. The `tfw_msg_write()` function is used only for preallocated skbs, and data size is known before skb allocations, thus `data` should perfectly fit the `it`. 

The reason for `WARN_ON` is invalid `it` state, where is `it->frag` points to nonexistent fragment.